### ISSUE TITLE
decimal support for agg function

### ISSUE THREAD

**Issue by liukun4515** [apache/arrow-datafusion#1545]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
The decimal data type has been supported in #122, but in this issue #122 we just support several agg functions. 
In this issue we will list some functions which are not implemented for the decimal data type.

- [ ] sum https://github.com/apache/arrow-datafusion/pull/1408

- [ ] min/max https://github.com/apache/arrow-datafusion/pull/1407

- [ ] count/DistinctCount: don't need to implement

- [ ] avg https://github.com/apache/arrow-datafusion/pull/1408

- [ ] var(col1): calculate the variance (sample) of col1

- [ ] var_samp(col1): calculate the variance (sample) of col1

- [ ] var_pop(col1): calculate the variance (population) of col1

- [ ] stddev(col1): calculate the standard deviation (sample) of col1

- [ ] stddev_samp(col1): calculate the standard deviation (sample) of col1

- [ ] stddev_pop(col1): calculate the standard deviation (population) of col1

TODO

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by liukun4515** [comment]
@realno I add this issue to track agg function which needs to be implemented decimal type.
You have added some agg functions in https://github.com/apache/arrow-datafusion/pull/1525, I have tracked them.

**Comment by realno** [comment]
@liukun4515 I have added more functions and am planning to add a few in the next few days. Here are the list, please add as tasks:

covar(col1, col2): calculate the covariance (sample) of col1 and col2

covar_samp(col1, col2): calculate the covariance (sample) of col1 and col2

covar_pop(col1, col2): calculate the covariance (population) of col1 and col2

corr(col1, col2): calculate the correlation coefficient (sample) of col1 and col2

corr_samp(col1, col2): calculate the  correlation coefficient (sample) of col1 and col2

corr_pop(col1, col2): calculate the  correlation coefficient (population) of col1 and col2

median(col1): calculate the median of col1

**Comment by liukun4515** [comment]
> @liukun4515 I have added more functions and am planning to add a few in the next few days. Here are the list, please add as tasks:
> 
> covar(col1, col2): calculate the covariance (sample) of col1 and col2
> 
> covar_samp(col1, col2): calculate the covariance (sample) of col1 and col2
> 
> covar_pop(col1, col2): calculate the covariance (population) of col1 and col2
> 
> corr(col1, col2): calculate the correlation coefficient (sample) of col1 and col2
> 
> corr_samp(col1, col2): calculate the correlation coefficient (sample) of col1 and col2
> 
> corr_pop(col1, col2): calculate the correlation coefficient (population) of col1 and col2
> 
> median(col1): calculate the median of col1

@realno The list of function are not agg function, and they are the builtin-function for datafusion.
Now the decimal is not supported in the builtin-function.
I will support them with decimal data when refactoring this model.

I will file a new issue to track builtin-function.

**Comment by liukun4515** [comment]
> e next few da

thanks @realno , I will add them.
You can add the agg functions first, the decimal supporting is not an emergency work.

**Comment by liukun4515** [comment]
> @liukun4515 I have added more functions and am planning to add a few in the next few days. Here are the list, please add as tasks:
> 
> covar(col1, col2): calculate the covariance (sample) of col1 and col2
> 
> covar_samp(col1, col2): calculate the covariance (sample) of col1 and col2
> 
> covar_pop(col1, col2): calculate the covariance (population) of col1 and col2
> 
> corr(col1, col2): calculate the correlation coefficient (sample) of col1 and col2
> 
> corr_samp(col1, col2): calculate the correlation coefficient (sample) of col1 and col2
> 
> corr_pop(col1, col2): calculate the correlation coefficient (population) of col1 and col2
> 
> median(col1): calculate the median of col1

@realno  please check the document of pg https://www.postgresql.org/docs/14/functions-aggregate.html
Some of agg functions don't need to support decimal type, and just need to support double precision data type(double/float64).

**Comment by kumarUjjawal** [comment]
@liukun4515 What's the status of this, do we still require the remaining work, I would like to take a look at remaining todos.

**Comment by Jefffrey** [comment]
> [@liukun4515](https://github.com/liukun4515) What's the status of this, do we still require the remaining work, I would like to take a look at remaining todos.

Feel free to take a look at which functions lack native decimal support and implement for them

**Comment by lgingerich** [comment]
I was taking a look into this and spent a bit of time on the algorithmic side of it to compare numerical accuracy of Welford+float, Welford+decimal, and sum-of-squares + decimal.

TLDR is that it seems that Welford+float is always better than attempting to calculate the variance (and therefore std dev) natively for decimals due to how truncation/rounding error accumulates.

Reference: https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm


Here's a simple example of variance error accumulation for a set of random inputs:
<img width="2401" height="1177" alt="Image" src="https://github.com/user-attachments/assets/41f67276-5e6d-498e-9e26-89dad5d36249" />

**Comment by theirix** [comment]
I believe two other similar issues #3481 and #3572 could be closed in favour of this more discussed issue
